crates/common/axum_tls/src/maybe_tls.rs

```diff
@@ -0,0 +1,65 @@
+use pin_project::pin_project;
+use std::io;
+use std::io::Error;
+use std::pin::Pin;
+use std::task::Poll;
+use tokio::io::AsyncRead;
+use tokio::io::AsyncWrite;
+use tokio::io::BufReader;
+use tokio::io::ReadBuf;
+use tokio_rustls::server::TlsStream;
+
+#[pin_project(project = MaybeTlsStreamProj)]
+/// An optional [TlsStream], i.e. a stream of either TLS or non-TLS data
+///
+/// This is useful for redirecting HTTP requests to HTTPS.
+pub enum MaybeTlsStream<I> {
```

Comment:
Kudos!

When I asked to redirect HTTP requests to HTTPS, I didn't imagine a sec such a dance with `pin` and streams.

Reply:
Yeah, the trait requirements for `hyper` and `axum` machinery are low level. But `pin_project` makes it almost trivial